s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/S3SourceTask.java

```diff
@@ -0,0 +1,240 @@
+
+    private final Object pollLock = new Object();
+    private FileReader fileReader;
+    private final Set<String> failedObjectKeys = new HashSet<>();
+    private final Set<String> inProcessObjectKeys = new HashSet<>();
+
+    private OffsetManager offsetManager;
+
+    @SuppressWarnings("PMD.UnnecessaryConstructor")
+    public S3SourceTask() {
+        super();
+    }
+
+    @Override
+    public String version() {
+        return Version.VERSION;
+    }
+
+    @Override
+    public void start(final Map<String, String> props) {
+        LOGGER.info("S3 Source task started.");
+        s3SourceConfig = new S3SourceConfig(props);
+        initializeConverters();
+        initializeS3Client();
+        this.s3Bucket = s3SourceConfig.getString(AWS_S3_BUCKET_NAME_CONFIG);
+        this.transformer = TransformerFactory.getTransformer(s3SourceConfig);
+        offsetManager = new OffsetManager(context, s3SourceConfig);
+        fileReader = new FileReader(s3SourceConfig, this.s3Bucket, failedObjectKeys);
+        prepareReaderFromOffsetStorageReader();
+        this.taskInitialized = true;
+    }
+
+    private void initializeConverters() {
+        try {
+            keyConverter = Optional
+                    .of((Converter) Class.forName((String) s3SourceConfig.originals().get("key.converter"))
+                            .getDeclaredConstructor()
+                            .newInstance());
+            valueConverter = (Converter) Class.forName((String) s3SourceConfig.originals().get("value.converter"))
+                    .getDeclaredConstructor()
+                    .newInstance();
+        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | InvocationTargetException
+                | NoSuchMethodException e) {
+            throw new ConnectException("Connect converters could not be instantiated.", e);
+        }
+    }
+
+    private void initializeS3Client() {
+        this.s3Client = s3ClientFactory.createAmazonS3Client(s3SourceConfig);
+        LOGGER.debug("S3 client initialized");
+    }
+
+    private void prepareReaderFromOffsetStorageReader() {
+        sourceRecordIterator = new SourceRecordIterator(s3SourceConfig, s3Client, this.s3Bucket, offsetManager,
+                this.transformer, fileReader);
+    }
+
+    @Override
+    public List<SourceRecord> poll() throws InterruptedException {
+        LOGGER.info("Polling again");
+        synchronized (pollLock) {
+            final List<SourceRecord> results = new ArrayList<>(s3SourceConfig.getInt(MAX_POLL_RECORDS));
+
+            if (connectorStopped.get()) {
+                return results;
+            }
+
+            while (!connectorStopped.get()) {
+                try {
+                    LOGGER.info("Number of records sent {}", extractSourceRecords(results).size());
+                    return results;
+                } catch (AmazonS3Exception | DataException exception) {
```

Comment:
This looks like it can be simplified by breaking out the DataException into its own block and allowing the catch to do the work that is being done in handleException.
e.g.
  } catch (DataException exception) {
                    LOGGER.warn("DataException. Will NOT try again.", exception);
                }